src/main/java/com/redhat/pantheon/model/api/SlingResource.java

```diff
@@ -0,0 +1,274 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.Arrays.stream;
+
+/**
+ * An Editable Sling resource model object. These are simple objects made up of a single constructor based on an existing
+ * sling resource, and a set of fields or child resources mapped to the resource. Fields and child resources are
+ * strongly typed and they can be modified. SlingResources are wrappers around existing sling resources and as such the
+ * resource needs to exist before it can be wrapped by one of these.
+ */
+public class SlingResource implements Adaptable {
+
+    private final Resource resource;
+
+    public SlingResource(@Nonnull Resource resource) {
+        this.resource = resource;
+    }
+
+    public Resource getResource() {
+        return resource;
+    }
+
+    /**
+     * Initializes all fields which have a default value set, and which aren't already set.
+     * This method is only meant to be called after new resources are created.
+     */
+    protected void initDefaultValues() {
+        getMembers(Field.class)
+                // discard fields which already have a value set
+                .filter(field -> !field.isSet())
+                // only consider fields which have a default value
+                .filter(field -> field.defaultValue.isPresent())
+                .forEach(field -> field.set(field.defaultValue.get()));
+    }
+
+    /**
+     * Returns all SlingModel members (assigned fields which implement the ResourceMember interdace
+     * @return
+     */
+    private Stream<ResourceMember> allMembers() {
+        return stream(this.getClass().getDeclaredFields())
+                // only fields which implement ResourceMember
+                .filter(reflectedField -> ResourceMember.class.isAssignableFrom(reflectedField.getType()))
+                // convert to the field values
+                .map(field -> {
+                    try {
+                        return (ResourceMember) field.get(SlingResource.this);
+                    } catch (IllegalAccessException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                // only initialized fields
+                .filter(field -> field != null);
+    }
+
+    /**
+     * Returns all SlingModel members of a specific implementation.
+     */
+    private <A extends ResourceMember> Stream<A> getMembers(@Nonnull Class<A> memberClass) {
+        return (Stream<A>) allMembers()
+                .filter(resourceMember -> memberClass.isAssignableFrom(resourceMember.getClass()));
+    }
+
+    /**
+     * Returns a map with all the fields (deep fields included) for this model. The keys for the map
+     * are the jcr field names.
+     *
+     * @param excluding A list of JCR field names to exclude from the returned map.
+     * @return
+     */
+    public Map<String, Object> toMap(String ... excluding) {
```

Comment:
What is the use case here? Why do we (as developers) want to go through the effort of specifying entries that we don't want in the map?

Reply:
Yes. One of the over-engineered parts I mentioned :smile: 

The use is precisely restricting the amount of data that we (for example) want to transmit via the API or eventually restrict because of permissions.

Reply:
Fair enough! Maybe over-engineered, but only a little bit. ;)